sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TrackNotImplementedException.cs

```diff
@@ -50,12 +51,32 @@ protected override void Initialize(SonarAnalysisContext context)
                         return;
                     }
 
-                    if (c.SemanticModel.GetTypeInfo(throwStatement.Expression).Type.Is(KnownType.System_NotImplementedException))
+                    ReportDiagnostic(c, throwStatement.Expression, throwStatement);
+                },
+                SyntaxKind.ThrowStatement);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var creationExpression = (ObjectCreationExpressionSyntax)c.Node;
+                    var throwExpression = creationExpression.Parent;
+
+                    if (throwExpression.Kind() != SyntaxKindEx.ThrowExpression)
                     {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, throwStatement.GetLocation()));
+                        return;
                     }
+
+                    ReportDiagnostic(c, creationExpression, throwExpression);
                 },
-                SyntaxKind.ThrowStatement);
+                SyntaxKind.ObjectCreationExpression);
```

Comment:
coming back on this one, @Corniel - why do you have to enter this callback on `ObjectCreationExpressionSyntax` and not on `SyntaxKindEx.ThrowExpression` ?

it looks cumbersome to always enter this callback for all object creation expression, then look at the `Parent` - is the TypeInfo only at the `creationExpression` level and not at the `throwExpression` level?

even so, it looks more intuitive to enter the callback only for `SyntaxKindEx.ThrowExpression` and then go inside it and get the `ObjectCreationExpression` from it to get the type..

Reply:
@andrei-epure-sonarsource Because the type you get back is a type that is not in our version. And the base class that is has no Expression. I considered using `dynamic`, but that is never used yet, and then I had to add an dependency on MS.CSharp. Also using reflection felt like the wrong way to go.

Reply:
tests are green with this version
```
            context.RegisterSyntaxNodeActionInNonGenerated(
                c =>
                {
                    var throwExpression = (ThrowExpressionSyntaxWrapper) c.Node;
                    ReportDiagnostic(c, throwExpression.Expression, throwExpression);
                },
                SyntaxKindEx.ThrowExpression);

```

and then I think you can simplify a lot the code , have a single callback

Reply:
Ah, you already created some nice reusable reflection based code.

Reply:
it's the Lightup layer in StyleCop.Analyzers - https://github.com/DotNetAnalyzers/StyleCopAnalyzers/tree/master/StyleCop.Analyzers/StyleCop.Analyzers/Lightup

we contributed to it, but credits go mainly to the StyleCop maintainers